core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -1030,8 +1034,22 @@ public String allocationId() {
         }
     }
 
+    /**
+     * The {@code ReplicasProxy} is an implementation of the {@code Replicas}
+     * interface that performs the actual {@code ReplicaRequest} on the replica
+     * shards. It also encapsulates the logic required for failing the replica
+     * if deemed necessary as well as marking it as stale when needed.
+     */
     final class ReplicasProxy implements ReplicationOperation.Replicas<ReplicaRequest> {
```

Comment:
@bleskes that's actually a scary thing, that "*works*" for Intellij's compiler (it reports no compilation errors), but if you actually run `gradle compileJava`:

```
/home/hinmanm/es/elasticsearch/core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java:347: error: name clash: failShardIfNeeded(ShardRouting,long,String,Exception,Runnable,Consumer<Exception>,Consumer<Exception>) in TransportWriteAction.WriteActionReplicasProxy and failShardIfNeeded(ShardRouting,long,String,Exception,Runnable,Consumer<Exception>,Consumer<Exception>) in TransportReplicationAction.ReplicasProxy have the same erasure, yet neither overrides the other
        public void failShardIfNeeded(ShardRouting replica, long primaryTerm, String message, Exception exception,
                    ^
/home/hinmanm/es/elasticsearch/core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java:346: error: method does not override or implement a method from a supertype
        @Override
        ^
```

It fails because of type erasure (I think due to the crazy generics that are in use), I tried a lot of different ways to get around it but I wasn't able to find a way.